pkg/ingester/metadata_map.go

```diff
@@ -0,0 +1,153 @@
+package ingester
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+)
+
+type metadataMap struct {
+	mtx              sync.Mutex
```

Comment:
Since this map is always accessed while holding `i.userStatesMtx` (not 100% sure, but it looks like it is to me), does it really need another two (!) levels of locking?

Reply:
Ditto from above about the locking.